test/functional/lua/filetype_spec.lua

```diff
@@ -70,6 +70,7 @@ describe('vim.filetype', function()
   it('works with patterns', function()
     eq('markdown', exec_lua([[
       local root = ...
+      vim.env.HOME = '/a-funky+home%dir'
```

Comment:
This current doesn't do anything because of #17501.

Reply:
If you revert the fix, the test fails, so it does seem to be working.  That might just be because the relevant code accesses `vim.env.HOME` directly as well.

Reply:
Oh, I was thinking about `expand()`.